yourbench/pipeline/answer_generation.py

```diff
@@ -0,0 +1,276 @@
+        return
+
+    # Run them
+    responses = run_inference(
+        config=config,
+        step_name="answer_generation",  # "model_roles" can link here
+        inference_calls=calls,
+    )
+    if not responses:
+        logger.warning(f"No responses returned for subset={source_subset}, scenario={scenario}.")
+        return
+
+    # Parse and assemble
+    final_ds = _parse_and_assemble(
+        question_ds=question_ds, responses_dict=responses, row_map=row_map, scenario=scenario
+    )
+    if final_ds is None or len(final_ds) == 0:
+        logger.warning(f"No answers parsed for subset={source_subset}, scenario={scenario}.")
+        return
+
+    # Append to the same subset
+    custom_save_dataset(dataset=final_ds, config=config, subset=output_subset)
+    logger.info(f"Appended {len(final_ds)} new answers to '{output_subset}' (scenario={scenario}).")
+
+
+def _build_inference_calls_scenario(config: Dict[str, Any], question_ds: Dataset, scenario: str, source_subset: str):
+    """
+    For each row in question_ds, build an InferenceCall with the appropriate prompt:
+     - "zero_shot" => use ZEROSHOT_QA_USER_PROMPT
+     - "with_correct_chunk" => retrieve chunk(s) from 'chunked' dataset and use GOLD_QA_USER_PROMPT
+
+    Returns (calls, row_map) for re-assembly after inference.
+    """
+    calls = []
+    row_map = []
+
+    if "question" not in question_ds.column_names:
+        logger.error("Dataset missing 'question' column. Cannot build calls.")
+        return calls, row_map
+
+    doc_meta_map = {}
+    if scenario == "with_correct_chunk":
+        # Load chunked dataset to retrieve correct chunk text
+        logger.info("Loading 'chunked' dataset to retrieve chunk text for with_correct_chunk scenario.")
+        try:
+            chunked_ds = custom_load_dataset(config, subset="chunked")
+            doc_meta_map = _build_doc_meta_map(chunked_ds)
+        except Exception as e:
+            logger.error(f"Failed loading chunked dataset for scenario={scenario}: {e}")
+            return calls, row_map
+
+    for i, row in enumerate(question_ds):
+        qtext = row.get("question", "")
+        if not qtext or not isinstance(qtext, str):
+            continue
+
+        if scenario == "zero_shot":
+            user_prompt = ZEROSHOT_QA_USER_PROMPT.format(question=qtext)
+            user_msg = {"role": "user", "content": user_prompt}
+            calls.append(InferenceCall(messages=[user_msg], tags=["zero_shot"]))
+            row_map.append(i)
+
+        elif scenario == "with_correct_chunk":
+            doc_id = row.get("document_id", "")
+            if not doc_id:
+                logger.debug(f"Row {i} missing document_id, skipping with_correct_chunk.")
+                continue
+
+            # Single-shot uses 'chunk_id', multi-hop uses 'source_chunk_ids'
+            if source_subset.startswith("single_shot"):
+                chunk_id = row.get("chunk_id", "")
```

Comment:
Ŵhy don't you use an extraction function here?

Reply:
added `_extract_chunk_ids_for_row(...)` which  pulls out document_id and chunk IDs for single-hop vs. multi-hop